src/openrct2/scenario/ObjectiveGoal.h

```diff
@@ -0,0 +1,647 @@
+        , minRideLengthGoal(_minLen)
+        , maxRideLengthGoal(_maxLen){};
+
+public:
+    ObjectiveTransportRidesGoal(
+        uint8_t _minNum, uint8_t _maxNum, uint16_t _minLen, uint16_t _maxLen, bool _unique, bool _finishedExistingRides)
+        : ObjectiveGentleRidesGoal(_minNum, _maxNum, _unique, _finishedExistingRides, GoalID::TransportRidesGoal)
+        , minRideLengthGoal(_minLen)
+        , maxRideLengthGoal(_maxLen){};
+    bool CheckCondition() override;
+    bool CheckSpecialRequirements(rct_string_id& error) const override;
+    uint16_t GetMinRideLengthGoal() const
+    {
+        return minRideLengthGoal;
+    }
+    void SetMinRideLengthGoal(uint16_t _minRideLengthGoal)
+    {
+        minRideLengthGoal = _minRideLengthGoal;
+        redoToString = true;
+    }
+    uint16_t GetMaxRideLengthGoal() const
+    {
+        return maxRideLengthGoal;
+    }
+    void SetMaxRideLengthGoal(uint16_t _maxRideLengthGoal)
+    {
+        maxRideLengthGoal = _maxRideLengthGoal;
+        redoToString = true;
+    }
+};
+
+/// <summary>
+/// Build a number of waterrides
+/// </summary>
+class ObjectiveWaterRidesGoal : public ObjectiveTransportRidesGoal
+{
+protected:
+    uint16_t minRideExcitementGoal;
+    uint16_t maxRideExcitementGoal;
+
+    ObjectiveWaterRidesGoal(
+        uint8_t _minNum, uint8_t _maxNum, uint16_t _minLen, uint16_t _maxLen, float _minEx, float _maxEx, bool _unique,
+        bool _finishedExistingRides, GoalID _id)
+        : ObjectiveTransportRidesGoal(_minNum, _maxNum, _minLen, _maxLen, _unique, _finishedExistingRides, _id)
+        , minRideExcitementGoal(uint16_t(_minEx * 100))
+        , maxRideExcitementGoal(uint16_t(_maxEx * 100)){};
+
+public:
+    ObjectiveWaterRidesGoal(
+        uint8_t _minNum, uint8_t _maxNum, uint16_t _minLen, uint16_t _maxLen, float _minEx, float _maxEx, bool _unique,
+        bool _finishedExistingRides)
+        : ObjectiveTransportRidesGoal(
+            _minNum, _maxNum, _minLen, _maxLen, _unique, _finishedExistingRides, GoalID::WaterRidesGoal)
+        , minRideExcitementGoal(uint16_t(_minEx * 100))
+        , maxRideExcitementGoal(uint16_t(_maxEx * 100)){};
+    bool CheckCondition() override;
+    bool CheckSpecialRequirements(rct_string_id& error) const override;
+    uint16_t GetMinRideExcitementGoal() const
+    {
+        return minRideExcitementGoal;
+    }
+    void SetMinRideExcitementGoal(uint16_t _minRideExcitementGoal)
+    {
+        minRideExcitementGoal = _minRideExcitementGoal;
+        redoToString = true;
+    }
+    uint16_t GetMaxRideExcitementGoal() const
+    {
+        return maxRideExcitementGoal;
+    }
+    void SetMaxRideExcitementGoal(uint16_t _maxRideExcitementGoal)
+    {
+        maxRideExcitementGoal = _maxRideExcitementGoal;
+        redoToString = true;
+    }
+};
+
+/// <summary>
+/// Build a number of roller coasters
+/// </summary>
+class ObjectiveCoasterGoal final : public ObjectiveWaterRidesGoal
```

Comment:
> No, but it does prevent needing to declare duplicate variables everywhere and saves a lot of lines especially later when it comes to displaying it.

Sure, but why not just have all the variables in a single class? Most of them can apply to almost all ride types.

Reply:
It's just a choice I made to prevent putting the logic at other places, where depending on what ride category it is, it will or will not allow certain values to be entered, and it will or will not need to check for them.